database/migrations/2024_10_20_154356_create_qos_table.php

```diff
@@ -0,0 +1,64 @@
+            $table->string('title')->comment('Graph Title');
+            $table->string('snmp_idx')->comment('SNMP Index for polling QoS data');
+            $table->string('rrd_id')->comment('Suffix for the RRD file to identify this QoS');
+            $table->boolean('ingress')->default(0)->comment('Does this process ingress bytes');
+            $table->boolean('egress')->default(0)->comment('Does this process egress bytes');
+            $table->boolean('disabled')->default(0)->comment('Should this QoS be polled');
+            $table->boolean('ignore')->default(0)->comment('Should this QoS be alerted on');
+            $table->bigInteger('last_polled')->nullable()->comment('Last polled time for calculating rate');
+            $table->bigInteger('max_in')->nullable()->comment('Maximum value for input data if defined');
+            $table->bigInteger('max_out')->nullable()->comment('Maximum value for output data if defined');
+            $table->bigInteger('last_bytes_in')->nullable()->comment('Last polled counter for input bytes');
+            $table->bigInteger('last_bytes_out')->nullable()->comment('Last polled counter for output bytes');
+            $table->bigInteger('bytes_in_rate')->nullable()->comment('Output rate for bytes');
+            $table->bigInteger('bytes_out_rate')->nullable()->comment('Input rate for bytes');
+            $table->bigInteger('last_bytes_drop_in')->nullable()->comment('Last polled counter for input dropped bytes');
+            $table->bigInteger('last_bytes_drop_out')->nullable()->comment('Last polled counter for output dropped bytes');
+            $table->bigInteger('bytes_drop_in_rate')->nullable()->comment('Output rate for dropped bytes');
+            $table->bigInteger('bytes_drop_out_rate')->nullable()->comment('Input rate for dropped bytes');
+            $table->bigInteger('last_packets_in')->nullable()->comment('Last polled counter for input packets');
+            $table->bigInteger('last_packets_out')->nullable()->comment('Last polled counter for output packets');
+            $table->bigInteger('packets_in_rate')->nullable()->comment('Output rate for packets');
+            $table->bigInteger('packets_out_rate')->nullable()->comment('Input rate for packets');
+            $table->bigInteger('last_packets_drop_in')->nullable()->comment('Last polled counter for input dropped packets');
+            $table->bigInteger('last_packets_drop_out')->nullable()->comment('Last polled counter for output dropped packets');
+            $table->bigInteger('packets_drop_in_rate')->nullable()->comment('Output rate for dropped packets');
+            $table->bigInteger('packets_drop_out_rate')->nullable()->comment('Input rate for dropped packets');
+            $table->bigInteger('bytes_drop_in_pct')->storedAs('CASE WHEN bytes_in_rate = 0 THEN 0 ELSE 100 * bytes_drop_in_rate / bytes_in_rate END')->comment('Percentage of input bytes dropped');
```

Comment:
I'm not sure about this.  Adding logic at the database layer has bitten us in the ass a lot.

Reply:
Logic has been moved to the observer code